File: lilac/src/test/java/com/hartwig/hmftools/lilac/evidence/CombineEvidenceTest.java
package com.hartwig.hmftools.lilac.evidence;

import static junit.framework.TestCase.assertFalse;
import static junit.framework.TestCase.assertTrue;

import java.util.Map;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;

import org.junit.Test;

public class CombineEvidenceTest
{
    @Test
    public void testCheckSingles()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("T", 13);
        evidence.put("S", 71);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(344), evidence);

        evidence = Maps.newHashMap();
        evidence.put("S", 20);
        evidence.put("C", 20);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(348), evidence);

        evidence = Maps.newHashMap();
        evidence.put("SS", 9);
        evidence.put("SC", 4);
        PhasedEvidence combined = new PhasedEvidence(Lists.newArrayList(344, 348), evidence);

        assertFalse(CombineEvidence.canCombine(left, combined, right));
    }

    @Test
    public void testRealLife()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AM", 36);
        evidence.put("LM", 16);
        evidence.put("RT", 11);
        evidence.put("RM", 35);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(1, 3), evidence);

        evidence = Maps.newHashMap();
        evidence.put("AMT", 31);
        evidence.put("AMA", 1);
        evidence.put("LMT", 15);
        evidence.put("RTT", 9);
        evidence.put("RTA", 1);
        evidence.put("RMA", 16);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(1,3,7), evidence);

        assertTrue(CombineEvidence.canCombine(left, right));
    }

    @Test
    public void canCombineExtraBaseToRight()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TT", 3);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);

        evidence = Maps.newHashMap();
        evidence.put("ARS", 2);
        evidence.put("TTH", 3);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(3, 7, 10), evidence);

        assertTrue(CombineEvidence.canCombine(left, right));
    }

    @Test
    public void cannotCombineExtraBaseToRight()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TT", 3);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);

        evidence = Maps.newHashMap();
        evidence.put("ARS", 2);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(3, 7, 10), evidence);
        assertFalse(CombineEvidence.canCombine(left, right));
    }

    @Test
    public void canCombineExtraBaseToLeft()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("HAR", 2);
        evidence.put("ATT", 3);

        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(2, 3, 7), evidence);
        evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TT", 3);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);
        assertTrue(CombineEvidence.canCombine(left, right));
    }

    @Test
    public void canCombineOverlappingAndMatchingButNotWhenReversed()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TT", 3);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);

        evidence = Maps.newHashMap();
        evidence.put("RS", 2);
        evidence.put("TS", 3);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(7, 10), evidence);
        assertTrue(CombineEvidence.canCombine(left, right));
        assertFalse(CombineEvidence.canCombine(right, left));
    }

    @Test
    public void cannotCombineWhenSequencesAreNotAccountedFor()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TT", 3);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);

        evidence = Maps.newHashMap();
        evidence.put("SS", 2);
        evidence.put("TS", 3);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(7, 1), evidence);
        assertFalse(CombineEvidence.canCombine(left, right));
    }

    @Test
    public void cannotCombineNonOverlapping()
    {
        Map<String,Integer> evidence = Maps.newHashMap();
        evidence.put("AR", 2);
        evidence.put("TR", 3);
        PhasedEvidence left = new PhasedEvidence(Lists.newArrayList(3, 7), evidence);

        evidence = Maps.newHashMap();
        evidence.put("TS", 2);
        evidence.put("QR", 3);
        PhasedEvidence right = new PhasedEvidence(Lists.newArrayList(9, 10), evidence);
        assertFalse(CombineEvidence.canCombine(left, right));
    }

}

File: lilac/src/test/kotlin/com/hartwig/hmftools/lilackt/evidence/CombineEvidenceTest.kt
package com.hartwig.hmftools.lilackt.evidence

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CombineEvidenceTest {

    @Test
    fun testCheckSingles() {
        val left = PhasedEvidence(listOf(344).toIntArray(), mapOf(Pair("T", 13), Pair("S", 71)))
        val right = PhasedEvidence(listOf(348).toIntArray(), mapOf(Pair("S", 20), Pair("C", 20)))
        val combined = PhasedEvidence(listOf(344, 348).toIntArray(), mapOf(Pair("SS", 9), Pair("SC", 4)))
        assertFalse(CombineEvidence.canCombine(left, combined, right))
    }

    @Test
    fun testRealLife() {
        val left = PhasedEvidence(listOf(1, 3).toIntArray(), mapOf(Pair("AM", 36), Pair("LM", 16),  Pair("RT", 11),  Pair("RM", 35)))
        val right = PhasedEvidence(listOf(1, 3, 7).toIntArray(), mapOf(Pair("AMT", 31), Pair("AMA", 1), Pair("LMT", 15),  Pair("RTT", 9),  Pair("RTA", 1),  Pair("RMA", 16)))
        assertTrue(CombineEvidence.canCombine(left, right))

    }

    @Test
    fun canCombineExtraBaseToRight() {
        val left = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TT", 3)))
        val right = PhasedEvidence(listOf(3, 7, 10).toIntArray(), mapOf(Pair("ARS", 2), Pair("TTH", 3)))
        assertTrue(CombineEvidence.canCombine(left, right))
    }

    @Test
    fun cannotCombineExtraBaseToRight() {
        val left = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TT", 3)))
        val right = PhasedEvidence(listOf(3, 7, 10).toIntArray(), mapOf(Pair("ARS", 2)))
        assertFalse(CombineEvidence.canCombine(left, right))
    }

    @Test
    fun canCombineExtraBaseToLeft() {
        val left = PhasedEvidence(listOf(2, 3, 7).toIntArray(), mapOf(Pair("HAR", 2), Pair("ATT", 3)))
        val right = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TT", 3)))
        assertTrue(CombineEvidence.canCombine(left, right))
    }

    @Test
    fun canCombineOverlappingAndMatchingButNotWhenReversed() {
        val left = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TT", 3)))
        val right = PhasedEvidence(listOf(7, 10).toIntArray(), mapOf(Pair("RS", 2), Pair("TS", 3)))
        assertTrue(CombineEvidence.canCombine(left, right))
        assertFalse(CombineEvidence.canCombine(right, left))
    }

    @Test
    fun cannotCombineWhenSequencesAreNotAccountedFor() {
        val left = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TT", 3)))
        val right = PhasedEvidence(listOf(7, 10).toIntArray(), mapOf(Pair("SS", 2), Pair("TS", 3)))
        assertFalse(CombineEvidence.canCombine(left, right))
    }

    @Test
    fun cannotCombineNonOverlapping() {
        val left = PhasedEvidence(listOf(3, 7).toIntArray(), mapOf(Pair("AR", 2), Pair("TR", 3)))
        val right = PhasedEvidence(listOf(9, 10).toIntArray(), mapOf(Pair("TS", 2), Pair("QR", 3)))
        assertFalse(CombineEvidence.canCombine(left, right))
    }

}